File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/xdnetplay/gen3/Gen3Text.kt
// SPDX-License-Identifier: GPL-2.0-or-later

package org.dolphinemu.dolphinemu.features.xdnetplay.gen3

/**
 * Gen 3 western (International) character set encode/decode.
 *
 * Ported 1:1 from the empirically validated Python reference (gen3save.py);
 * all byte values were confirmed against real Emerald saves (VALIDATION.md).
 * Strings are fixed-width, 0xFF-terminated; a string that fills the whole
 * field carries no terminator, exactly like the games.
 */
object Gen3Text {
    const val TERMINATOR = 0xFF

    private val byteToChar = HashMap<Int, Char>()
    private val charToByte = HashMap<Char, Int>()

    private fun map(b: Int, c: Char) {
        byteToChar[b] = c
        charToByte[c] = b
    }

    init {
        map(0x00, ' ')
        // Digits 0-9: 0xA1..0xAA
        for (i in 0..9) {
            map(0xA1 + i, '0' + i)
        }
        map(0xAB, '!')
        map(0xAC, '?')
        map(0xAD, '.')
        map(0xAE, '-')
        map(0xAF, '・')
        map(0xB0, '…')
        map(0xB1, '“')
        map(0xB2, '”')
        map(0xB3, '‘')
        map(0xB4, '’')
        map(0xB5, '♂')
        map(0xB6, '♀')
        map(0xB7, '$')
        map(0xB8, ',')
        map(0xB9, '×')
        map(0xBA, '/')
        // A-Z: 0xBB..0xD4, a-z: 0xD5..0xEE
        for (i in 0..25) {
            map(0xBB + i, 'A' + i)
            map(0xD5 + i, 'a' + i)
        }
    }

    /** True if [c] has a Gen 3 byte value in this charset. */
    fun canEncode(c: Char): Boolean = charToByte.containsKey(c)

    /**
     * Decode up to [maxLen] bytes of Gen 3 text starting at [off].
     * Unknown bytes decode as '?', mirroring the Python reference.
     */
    fun decode(buf: ByteArray, off: Int, maxLen: Int): String {
        val sb = StringBuilder()
        for (i in 0 until maxLen) {
            // Kotlin Byte is signed: mask to 0..255 before any comparison.
            val b = buf[off + i].toInt() and 0xFF
            if (b == TERMINATOR) {
                break
            }
            sb.append(byteToChar[b] ?: '?')
        }
        return sb.toString()
    }

    /**
     * Encode [text] to exactly [maxLen] bytes, 0xFF-terminated/padded.
     * Text longer than [maxLen] is truncated (as in the Python reference).
     *
     * @throws IllegalArgumentException if a character cannot be encoded.
     */
    fun encode(text: String, maxLen: Int): ByteArray {
        val out = ByteArray(maxLen) { TERMINATOR.toByte() }
        val n = minOf(text.length, maxLen)
        for (i in 0 until n) {
            val ch = text[i]
            val b = charToByte[ch]
                ?: throw IllegalArgumentException(
                    "character '$ch' not encodable in Gen 3 charset"
                )
            out[i] = b.toByte()
        }
        return out
    }

    /**
     * Make [text] Gen 3-encodable by force: straight ASCII quotes become the
     * charset's curly forms, anything else unencodable is DROPPED, and the
     * result is clamped to [maxLen] characters (0 = no clamp).
     *
     * Use this for input that must degrade rather than fail (a nickname out of
     * a Showdown paste, a trainer name that arrived over the wire); use
     * [encode]'s exception when a human is typing and deserves to be told which
     * character is the problem.
     */
    fun sanitize(text: String, maxLen: Int = 0): String {
        val sb = StringBuilder()
        for (c in text) {
            if (maxLen != 0 && sb.length == maxLen) {
                break
            }
            val mapped = when (c) {
                '\'' -> '’'
                '"' -> '”'
                else -> c
            }
            if (canEncode(mapped)) {
                sb.append(mapped)
            }
        }
        return sb.toString()
    }
}

File: Source/Core/UICommon/XDNetplay/Gen3Text.cpp
// Copyright 2026 Dolphin Emulator Project
// SPDX-License-Identifier: GPL-2.0-or-later

#include "UICommon/XDNetplay/Gen3Text.h"

#include <algorithm>
#include <map>

namespace XDNetplay::Gen3Text
{
namespace
{
struct Maps
{
  std::map<u8, char32_t> byte_to_char;
  std::map<char32_t, u8> char_to_byte;

  void Map(u8 b, char32_t c)
  {
    byte_to_char[b] = c;
    char_to_byte[c] = b;
  }

  Maps()
  {
    Map(0x00, U' ');
    // Digits 0-9: 0xA1..0xAA
    for (int i = 0; i <= 9; i++)
      Map(static_cast<u8>(0xA1 + i), U'0' + i);
    Map(0xAB, U'!');
    Map(0xAC, U'?');
    Map(0xAD, U'.');
    Map(0xAE, U'-');
    Map(0xAF, U'・');  // KATAKANA MIDDLE DOT
    Map(0xB0, U'…');  // HORIZONTAL ELLIPSIS
    Map(0xB1, U'“');  // LEFT DOUBLE QUOTATION MARK
    Map(0xB2, U'”');  // RIGHT DOUBLE QUOTATION MARK
    Map(0xB3, U'‘');  // LEFT SINGLE QUOTATION MARK
    Map(0xB4, U'’');  // RIGHT SINGLE QUOTATION MARK
    Map(0xB5, U'♂');  // MALE SIGN
    Map(0xB6, U'♀');  // FEMALE SIGN
    Map(0xB7, U'$');
    Map(0xB8, U',');
    Map(0xB9, U'×');  // MULTIPLICATION SIGN
    Map(0xBA, U'/');
    // A-Z: 0xBB..0xD4, a-z: 0xD5..0xEE
    for (int i = 0; i <= 25; i++)
    {
      Map(static_cast<u8>(0xBB + i), U'A' + i);
      Map(static_cast<u8>(0xD5 + i), U'a' + i);
    }
  }
};

const Maps& GetMaps()
{
  static const Maps maps;
  return maps;
}
}  // namespace

std::u32string DecodeUtf8(const std::string& text)
{
  std::u32string out;
  size_t i = 0;
  const size_t n = text.size();
  while (i < n)
  {
    const u8 b0 = static_cast<u8>(text[i]);
    char32_t cp = U'?';
    size_t len = 1;
    if (b0 < 0x80)
    {
      cp = b0;
    }
    else if ((b0 & 0xE0) == 0xC0 && i + 1 < n)
    {
      cp = ((b0 & 0x1F) << 6) | (static_cast<u8>(text[i + 1]) & 0x3F);
      len = 2;
    }
    else if ((b0 & 0xF0) == 0xE0 && i + 2 < n)
    {
      cp = ((b0 & 0x0F) << 12) | ((static_cast<u8>(text[i + 1]) & 0x3F) << 6) |
           (static_cast<u8>(text[i + 2]) & 0x3F);
      len = 3;
    }
    else if ((b0 & 0xF8) == 0xF0 && i + 3 < n)
    {
      cp = ((b0 & 0x07) << 18) | ((static_cast<u8>(text[i + 1]) & 0x3F) << 12) |
           ((static_cast<u8>(text[i + 2]) & 0x3F) << 6) | (static_cast<u8>(text[i + 3]) & 0x3F);
      len = 4;
    }
    out.push_back(cp);
    i += len;
  }
  return out;
}

std::string EncodeUtf8(const std::u32string& text)
{
  std::string out;
  for (const char32_t cp : text)
  {
    if (cp < 0x80)
    {
      out.push_back(static_cast<char>(cp));
    }
    else if (cp < 0x800)
    {
      out.push_back(static_cast<char>(0xC0 | (cp >> 6)));
      out.push_back(static_cast<char>(0x80 | (cp & 0x3F)));
    }
    else if (cp < 0x10000)
    {
      out.push_back(static_cast<char>(0xE0 | (cp >> 12)));
      out.push_back(static_cast<char>(0x80 | ((cp >> 6) & 0x3F)));
      out.push_back(static_cast<char>(0x80 | (cp & 0x3F)));
    }
    else
    {
      out.push_back(static_cast<char>(0xF0 | (cp >> 18)));
      out.push_back(static_cast<char>(0x80 | ((cp >> 12) & 0x3F)));
      out.push_back(static_cast<char>(0x80 | ((cp >> 6) & 0x3F)));
      out.push_back(static_cast<char>(0x80 | (cp & 0x3F)));
    }
  }
  return out;
}

bool CanEncode(char32_t c)
{
  return GetMaps().char_to_byte.count(c) != 0;
}

std::string Decode(const u8* buf, size_t max_len)
{
  const auto& maps = GetMaps();
  std::u32string out;
  for (size_t i = 0; i < max_len; i++)
  {
    const u8 b = buf[i];
    if (b == TERMINATOR)
      break;
    const auto it = maps.byte_to_char.find(b);
    out.push_back(it != maps.byte_to_char.end() ? it->second : U'?');
  }
  return EncodeUtf8(out);
}

std::optional<std::vector<u8>> Encode(const std::string& text, size_t max_len, std::string* error)
{
  const auto& maps = GetMaps();
  std::vector<u8> out(max_len, TERMINATOR);
  const std::u32string codepoints = DecodeUtf8(text);
  const size_t n = std::min(codepoints.size(), max_len);
  for (size_t i = 0; i < n; i++)
  {
    const auto it = maps.char_to_byte.find(codepoints[i]);
    if (it == maps.char_to_byte.end())
    {
      if (error)
        *error = "character '" + EncodeUtf8(std::u32string(1, codepoints[i])) +
                 "' not encodable in Gen 3 charset";
      return std::nullopt;
    }
    out[i] = it->second;
  }
  return out;
}

std::string Sanitize(const std::string& text, size_t max_len)
{
  std::u32string out;
  for (const char32_t c : DecodeUtf8(text))
  {
    if (max_len != 0 && out.size() == max_len)
      break;
    char32_t mapped = c;
    if (c == U'\'')
      mapped = U'’';
    else if (c == U'"')
      mapped = U'”';
    if (CanEncode(mapped))
      out.push_back(mapped);
  }
  return EncodeUtf8(out);
}
}  // namespace XDNetplay::Gen3Text
